### PR TITLE
Use np.linalg.multi_dot instead of multiple np.dot routines

### PR DESCRIPTION
**Pull request by maldil**
#### Reference issue
Not a reported issue

#### What does this implement/fix?
It is more succinct and effective to refactor code to use `np.linalg.multi_dot` rather than numerous `np.dot` routines.
What do you think about this change?

### CODE REVIEW

**Comment by cbrnr** [comment]
Looks good to me!

**Comment by agramfort** [comment]
I am a bit relunctant to do these changes automatically. In the specific use cases here we know size of the matrices so we put the () at the right place so the order of the dots are optimal. To do these changes I would suggest you time in a real use case eg taking the examples that use these methods on real data and see how it affects the run time. Does it sound reasonable?

**Comment by drammock** [comment]
> In the specific use cases here we know size of the matrices so we put the () at the right place so the order of the dots are optimal.

According to the docstring: "multi_dot chains numpy.dot and uses optimal parenthesization of the matrices".  So we can probably trust that it won't slow things down, but I agree with @agramfort that it's worth a quick timing to make sure.  I do think it's more readable than what we have now, although I'd argue it's *less* readable than using `@`:

```py
CMinvG = np.dot(np.dot(U / (S + eps), U.T), G)  # current; hard to read
CMinvG = np.linalg.multi_dot([U / (S + eps), U.T, G])  # better
CMinvG = U / (S + eps) @ U.T @ G  # most readable?
```

File: mne/decoding/tests/test_csp.py [hunk 1]
```diff
@@ -271,7 +271,7 @@ def test_ajd():
     A /= np.atleast_2d(np.sqrt(np.sum(A ** 2, 1))).T
     covmats = np.empty((n_times, n_channels, n_channels))
     for i in range(n_times):
-        covmats[i] = np.dot(np.dot(A, np.diag(diags[i])), A.T)
+        covmats[i] = np.linalg.multi_dot([A, np.diag(diags[i]), A.T])
```
Comment (larsoner) [inline mne/decoding/tests/test_csp.py:274]: If we're going to optimize for performance anyway, might as well do this
```suggestion
        np.linalg.multi_dot([A, np.diag(diags[i]), A.T], out=convmats[i])
```

**Review by larsoner — COMMENTED** [review]
I would be very surprised if NumPy's `multi_dot` led to a performance decrease for us, but it's probably worth some quick checks with arrays of these shapes to see if it's at least as good as what we had before. Let me know if you need help with this @maldil 

I agree with @drammock that `@` is the most readable, but it's probably not optimal from an efficiency standpoint, so +1 for moving to `multi_dot` when reasonable to do so.

**Comment by cbrnr** [comment]
If we're already benchmarking, if it turns out that `@` is not significantly slower, I'd vote for that option because it's the best in terms of readability.

**Comment by cbrnr** [comment]
Likely, but it would be better to check.

**Comment by drammock** [comment]
> we discovered that multiple `dot` calls perform well for square matrixes and `multi_dot` perform better for non-square matrixes. @ does not work well in either case.

Define what "works well" / "performs well" means here?  Faster?  Uses less memory?  Less prone to programmer error?  More numerically stable?  Something else?

**Comment by drammock** [comment]
> This is about time complexity

ok, well, quoting @larsoner:

> it's probably worth some quick checks with arrays of these shapes to see if it's at least as good as what we had before. Let me know if you need help with this @maldil

So the next step is figure out what shapes these matrices have, and run some timing tests.

**Comment by agramfort** [comment]
this was brought by @maldil on the pyRiemann repo and see a quick benchmark I did:

Screenshot 2022-07-06 at 09 49 03

in other words it's not a silver bullet for all use cases

**Comment by larsoner** [comment]
Do you need to update `sample` or `testing`? Something like this should work
```
mne.datasets.sample.data_path(force_update=True, verbose=True)
```

**Comment by larsoner** [comment]
You need to `pip install pytest-cov` or `conda install pytest-cov`

**Comment by larsoner** [comment]
> As another point, using the np.dot is preferable if you are certain of the matrix forms and the best ordering. If you are not, it is recommended that you use multi_dot, which will select the appropriate ordering for you. The matrix shapes for the aforementioned scenarios, in my opinion, are not pre-known and are instead dependent on user inputs.

Indeed for at least some of these it will depend on 1) the number of channels and 2) either the number of sources, or the number of PCA components retained. And these won't be fully caught by tests, since those often use many more channels and/or sources. But the fact that they speed up some of these already is sufficient for me at least to assume it'll generalize to the denser cases (where overhead is even less of a concern)!

My vote is to change all but case (4) to use `multi_dot` for now. We can look deeper, but based on the variable naming for (4) these all seem like they will be square, so for that I would just chain `@` for readability.

There are probably other parts of the codebase that could benefit from this as well based on a `git grep "np.\dot.*np\.dot"`, (which will itself only catch the instances that occur in a single line, and also won't catch any chained `@` calls). But we should probably tackle those in another PR with appropriate benchmarking, too.

File: mne/decoding/tests/test_csp.py (hunk 1 above)
Comment (larsoner) [reply mne/decoding/tests/test_csp.py:274]: Those requirements should be satisfied. At worst you'll need to add a dtype to the empty call but I doubt it

**Comment by cbrnr** [comment]
On my computer, `multi_dot` always performs worst, and `@` best. In addition, `@` is also best in readability, so my vote goes to replacing everything with `@`.

```
# multi_dot, dot, @
# Case 1:
11.1 µs ± 8.09 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
4.35 µs ± 2.39 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
3.37 µs ± 802 ns per loop (mean ± std. dev. of 7 runs, 10 loops each)

# Case 2:
10.6 ms ± 3.79 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
6.06 ms ± 1.42 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
3.91 ms ± 1.02 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

# Case 3:
12.6 µs ± 4.08 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
3.39 µs ± 1.35 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
3.26 µs ± 730 ns per loop (mean ± std. dev. of 7 runs, 10 loops each)

# Case 4:
9.9 µs ± 3.01 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
6.31 µs ± 6.51 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
3.6 µs ± 800 ns per loop (mean ± std. dev. of 7 runs, 10 loops each)

# Case 5:
40.5 µs ± 6.25 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
30.9 µs ± 642 ns per loop (mean ± std. dev. of 7 runs, 10 loops each)
30.4 µs ± 1.03 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

**Comment by hoechenberger** [comment]
> On my computer, `multi_dot` always performs worst, and `@` best. In addition, `@` is also best in readability, so my vote goes to replacing everything with `@`.

On my M1 machine, using your benchmark code, `multi_dot` is always the slowest, and `@` is slightly slower than `dot`, but probably not so much as to warrant the bad readability in comparison to `@`

Screen Shot 2022-07-26 at 10 15 21

**Comment by cbrnr** [comment]
OT: I'm a bit disappointed. Your M1 is only about 2x faster than my 8 year old MacBook Pro:

```
                                   multi_dot       dot         @
shapes                                                          
[(4, 4), (4, 1), (1, 4)]            0.000356  0.000034  0.000073
[(58, 58), (58, 58), (58, 23784)]   0.082977  0.067537  0.042829
[(1, 4), (4, 4), (4, 1)]            0.000283  0.000062  0.000097
[(4, 4), (4, 4), (4, 4)]            0.000197  0.000061  0.000052
[(38, 38), (38, 38), (38, 123)]     0.000381  0.000306  0.000320
```

But yes, I'd say we go for `@` here.

**Comment by hoechenberger** [comment]
> I'm a bit disappointed. Your M1 is only about 2x faster than my 8 year old MacBook Pro:

I have the slowest M1 and tried this while running on battery and with OpenBLAS. I believe there's lots of room for improvement.

**Comment by cbrnr** [comment]
With everything I've read about the M1, it should not throttle at all while on battery, and even the slowest one should be much faster than an 8 year old Intel chip. And yes, these are pretty synthetic and specialized benchmarks, but it still surprised me that the difference is so small. Maybe we have to try with much larger matrices. Or there might be something wrong with your Python setup (OpenBLAS is normally very fast, comparable to Accelerate).

**Comment by drammock** [comment]
Using @cbrnr's script, but changing `number` from 10 to 100, I get this:

```
                                   multi_dot       dot         @
shapes                                                          
[(4, 4), (4, 1), (1, 4)]            0.000599  0.000166  0.000178
[(58, 58), (58, 58), (58, 23784)]   0.221222  0.189372  0.164674
[(1, 4), (4, 4), (4, 1)]            0.000582  0.000242  0.000281
[(4, 4), (4, 4), (4, 4)]            0.000611  0.000272  0.000320
[(38, 38), (38, 38), (38, 123)]     0.003133  0.002683  0.002686
```

So on my system too, `multi_dot` is slowest in every case. 🤔

**Comment by cbrnr** [comment]
Maybe something to report upstream? I thought the only reason why `multi_dot` exists is because it is faster...

**Comment by drammock** [comment]
> Maybe something to report upstream? I thought the only reason why `multi_dot` exists is because it is faster...

I don't think so.  I think it's faster *if you're careless about putting in parentheses where it's sensible to do so*.  None of these test cases involve multiplication where the shapes would cause temporary creation of an intermediate larger matrix, or where the largest matrix is multiplied first (rather than last). Changing the order of parentheses like 

```
t2 = timeit("dot(A, dot(B, C))", number=n, globals=globals())
```

does indeed make `dot` much slower than `multidot` for such cases:

```
[(58, 58), (58, 58), (58, 23784)]   0.193553  0.654028  0.174059
```

So for me, `multi_dot` works as expected but is not too useful for the kinds of operations we do...  sure, we don't always know the dimensions in advance, but in an `N×M @ M×M @ M×P` operation we *usually* know that, for example, `N << M`, which is often enough info to place the parentheses optimally.

**Comment by larsoner** [comment]
> we don't always know the dimensions in advance, but in an N×M @ M×M @ M×P operation we usually know that, for example, N << M, which is often enough info to place the parentheses optimally.

I agree with this in general. I'd say let's

1. transition to preferring `@` instead of `np.dot` in new code, and when we touch old code
2. use `multi_dot` for cases where the shapes are variable enough (in the right ways) to benefit from it choosing the optimal order for us

And for (2) I don't think we've seen a compelling case for it yet. So my vote would be to close this PR at the moment. WDYT @maldil are you convinced?

**Comment by cbrnr** [comment]
Just a quick final question: what is the optimal order? Why do large matrices have to come last and what makes this faster than any other order?

**Comment by larsoner** [comment]
It's not really large vs small, it's tall vs short. If everything is square it doesn't matter.

Imagine an operation of the following, using shapes in place of var names: `(4, 1) @ (1, 4) @ (4, 4)`. If the order of operations is:

1. `((4, 1) @ (1, 4)) @ (4, 4)`: evaluating the first/left requires 16 multiplications to get to the next operation `(4, 4) @ (4, 4)`, and this requires 64 multiplications to get to the final output of shape `(4, 4)`, for a total of 80 multiplies
2. `(4, 1) @ ((1, 4) @ (4, 4))` evaluating the first/right requires 16 multiplications to get to the next operation `(4, 1) @ (1, 4)`, and this requires 16 operations to get to the final output of shape `(4, 4)`, for a total of 32 total multiplies

Here we're just talking about multiplies but the additions required are similar IIRC

**Comment by cbrnr** [comment]
Nice, thanks @larsoner!

**Comment by larsoner** [comment]
I'll close for now but if we find places with consistent speedups let's implement them. Thanks for pushing the conversation on this @maldil !

**Pull request closed without merging**